Add Enumerable to the Reader

x4d3 opened the pull request:
As the reader is already implementing `.each`, it would be a shame to not be an `Enumerable` 

https://apidock.com/ruby/Enumerable

x4d3 commented:
Dear @keithdoggett

Could you please review :)

keithdoggett commented:
Looks good to me. Do you think there will be any issues if a file is rewound?

BuonOmo reviewed (changes requested):
Thanks for the neat idea @x4d3, unfortunately the current behaviour makes it impossible... Since it is not idempotent, you could have something like:

![image](https://user-images.githubusercontent.com/11378424/99943732-0c406300-2d72-11eb-974b-f37a4b6285f9.png)

Or worst, `first(n)` will not return the same result every time.

![image](https://user-images.githubusercontent.com/11378424/99943925-5cb7c080-2d72-11eb-94c3-996c54862a76.png)

Hence I'd be ok with that PR if and only if you could make `each` idempotent :/

In that case however, I think the PR will be a great improvement to the current API Reader provides, since `each` is usually expected to be idempotent

x4d3 commented:
@BuonOmo 
Very good point and very true.

I've inspired myself from this [article](https://blog.appsignal.com/2018/05/29/ruby-magic-enumerable-and-enumerator.html) to do it properly.

And I've made `each` idempotent, by resetting the indexes. Let me know what you think ?

BuonOmo commented on lib/rgeo/shapefile/reader.rb:
```diff
@@ -342,7 +343,10 @@ def next
       # and yield the Reader::Record for each one.
 
       def each
+        return to_enum(:each) unless block_given?
```
An enumerator must have a definition of it's on size, otherwise, `enumurator.size == nil`. I think here it would be pretty trivial:

```suggestion
        return to_enum(:each) { @num_records } unless block_given?
```

BuonOmo commented on lib/rgeo/shapefile/reader.rb:
```diff
@@ -342,7 +343,10 @@ def next
 
       def each
+        return to_enum(:each) unless block_given?
         return unless @opened
```
Here I guess we should raise then, otherwise this would lead to a very weird behaviour.

BuonOmo commented on lib/rgeo/shapefile/reader.rb:
```diff
@@ -342,7 +343,10 @@ def next
       def each
+        return to_enum(:each) unless block_given?
         return unless @opened
+        # Each needs to be idempotent, therefore we reset all the internal indexes to their original value
+        rewind
         yield _read_next_record while @cur_record_index < @num_records
```
Since a user may now call every enumerable method, and still use the original `next` based api (which might be deprecated then ? @keithdoggett). I think we should ensure that `each` will keep users current index :

```suggestion
        # Each needs to be idempotent, therefore we reset all the internal indexes to their original value
        current_record_index = @cur_record_index
        rewind
        yield _read_next_record while @cur_record_index < @num_records
      ensure
        current_record_index && @cur_record_index = current_record_index
```

x4d3 replied:
Shoult not  it be
`current_record_index && seek_index(current_record_index)` 
?

BuonOmo replied:
Definitely! Moreover, I don't know what is the behavior when `seek_index` returns `false` but it may mean that it is not ok to continue then

BuonOmo reviewed (commented):
That is much better, and a very fast response as well ! I've added a few more comments to make it even more complete. @keithdoggett I'll let you review again as well after that I guess :)

BuonOmo commented on lib/rgeo/shapefile/reader.rb:
```diff
@@ -343,11 +343,14 @@ def next
         rewind
         yield _read_next_record while @cur_record_index < @num_records
+      ensure
+        current_record_index && @cur_record_index = current_record_index
```
Sorry i definitely made a mistake (i think you'll see it when testing). It should be 
```suggestion
        current_record_index && seek_index(current_record_index)
```

Otherwise we're not updating the file index!


Also @keithdoggett I'm not sure what the behavior is when the `seek_index` function returns false...

x4d3 replied:
I've actually went with
```
      def each
        return to_enum(:each) { @num_records } unless block_given?
        raise "File was not open" unless @opened
        # Each needs to be idempotent, therefore we reset all the internal indexes to their original value
        current_record_index = @cur_record_index
        rewind
        begin
          yield _read_next_record while @cur_record_index < @num_records
        ensure
          seek_index(current_record_index)
        end
      end
```

x4d3 commented:
@BuonOmo Thanks for all your comments.
I've addressed them, and added loads of test to make sure everything is working according to what we expect.
I also took the liberty of removing the `rewind` of the README as it's not necessary anymore.

BuonOmo commented on lib/rgeo/shapefile/reader.rb:
```diff
@@ -340,12 +342,20 @@ def next
-        return unless @opened
-        yield _read_next_record while @cur_record_index < @num_records
+        return to_enum(:each) { @num_records } unless block_given?
+        raise "File was not open" unless @opened
```
```suggestion
        raise IOError, "File was not open" unless @opened
```

BuonOmo commented on lib/rgeo/shapefile/reader.rb:
```diff
@@ -340,12 +342,20 @@ def next
+        raise "File was not open" unless @opened
+        # Each needs to be idempotent, therefore we reset all the internal indexes to their original value
+        current_record_index = @cur_record_index
+        rewind
```
Your begin block is much more clear than my ensure indeed, however, I guess this `rewind` should be part of it

BuonOmo reviewed (commented):
Two tiny comments, and I'll still wait a response from @keithdoggett about behavior of seek_index. Thanks again for the contribution

x4d3 commented:
@BuonOmo 
`each` with block is now returning the reader itself, to be the same as the `Array.each` behaviour (https://apidock.com/ruby/Array/each)
I don't think we need to worry about what `seek_index` returns anymore ?

BuonOmo commented:
@x4d3 my issue wasn't that you would return or false, even though it is a really good point to return `self` indeed. But is is rather that when `seek_info` returns `false` it looks like it means that it couldn't move to the requested position...

keithdoggett commented:
I'm not sure exactly what to do if `seek_index` is `false`. Is it possible to have `@cur_record_index` be larger than `@num_records` by using the pubic API?

Calling `rewind` if `seek_index` is `false` makes sense to me, but we may want to investigate how we could wind up in that state in the first place.

x4d3 commented:
I don't actually think it's possible with the public methods to have `@cur_record_index` greater than `@num_records`.

I think `seek_index` return false only if an external call is made with `index` greater than `@num_records`.

In our case, `current_record_index` will always be smaller than `@num_records` and therefore `seek_index` won't return false.
If `current_record_index` is greater than `@num_records` it means than the `Reader` was already in a degenerate state which is not related to the `.each` call.

BuonOmo reviewed (approved):
You convinced me ! Thanks again for the PR

@keithdoggett shall I let you merge and plan for the next update ? (Idk if we have something to merge other than this pr)

And I guess we should add a part to say 'write your contribution to history' in our contribution guidelines, I find it better to let people write it rather than doing this on version bump.

keithdoggett commented:
@x4d3 can you update the `History.md` file with this change under the ## Current header, then I'll merge it. Thanks!

x4d3 commented:
@keithdoggett done
